tests/e2e/preferential_topology_snapshot.go

```diff
@@ -0,0 +1,1285 @@
+		preferredDatastorePaths        []string
+		allowedTopologyRacks           []string
+		allowedTopologyForRack1        []v1.TopologySelectorLabelRequirement
+		allowedTopologyForRack2        []v1.TopologySelectorLabelRequirement
+		allowedTopologyForRack3        []v1.TopologySelectorLabelRequirement
+		rack1DatastoreListMap          map[string]string
+		rack2DatastoreListMap          map[string]string
+		rack3DatastoreListMap          map[string]string
+		storageclass                   *storagev1.StorageClass
+		err                            error
+		snapshotContentCreated         bool
+		snapshotCreated                bool
+		pandoraSyncWaitTime            int
+	)
+
+	ginkgo.BeforeEach(func() {
+		bootstrap()
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		client = f.ClientSet
+		namespace = getNamespaceToRunTests(f)
+		nodeList, err := fnodes.GetReadySchedulableNodes(f.ClientSet)
+		framework.ExpectNoError(err, "Unable to find ready and schedulable Node")
+		if !(len(nodeList.Items) > 0) {
+			framework.Failf("Unable to find ready and schedulable Node")
+		}
+
+		//Get snapshot client using the rest config
+		restConfig = getRestConfigClient()
+		snapc, err = snapclient.NewForConfig(restConfig)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		if os.Getenv(envPandoraSyncWaitTime) != "" {
+			pandoraSyncWaitTime, err = strconv.Atoi(os.Getenv(envPandoraSyncWaitTime))
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		} else {
+			pandoraSyncWaitTime = defaultPandoraSyncWaitTime
+		}
+
+		topologyLength, leafNode, leafNodeTag0, leafNodeTag1, leafNodeTag2 = 5, 4, 0, 1, 2
+		topologyMap := GetAndExpectStringEnvVar(topologyMap)
+		topologyAffinityDetails, topologyCategories = createTopologyMapLevel5(topologyMap,
+			topologyLength)
+		allowedTopologies = createAllowedTopolgies(topologyMap, topologyLength)
+
+		// fetching k8s master ip
+		allMasterIps = getK8sMasterIPs(ctx, client)
+		masterIp = allMasterIps[0]
+
+		// fetching datacenter details
+		dataCenters, err = e2eVSphere.getAllDatacenters(ctx)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// fetching cluster details
+		clusters, err = getTopologyLevel5ClusterGroupNames(masterIp, dataCenters)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		topologyAffinityDetails, topologyCategories = createTopologyMapLevel5(topologyMap,
+			topologyLength)
+		allowedTopologies = createAllowedTopolgies(topologyMap, topologyLength)
+
+		// fetching list of datatstores shared between vm's
+		shareddatastoreListMap, err = getListOfSharedDatastoresBetweenVMs(masterIp, dataCenters)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// fetching list of datastores available in different racks
+		rack1DatastoreListMap, err = getListOfDatastoresByClusterName(masterIp, clusters[0])
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		rack2DatastoreListMap, err = getListOfDatastoresByClusterName(masterIp, clusters[1])
```

Comment:
please put in a loop wherever possible

Reply:
This I will handle in a new PR as it will require lot of code change in entire preferential automation submitted so far and will require full E2E testing again for the new changes which will going to take sometime.
I have already raised one bug on preferential automation which I need to take care. I will handle your loop review comment in that bug.